WIP: Patterndb v6 to contain type hints

bazsi opened the pull request:
Fixing a compatibility issue with typing and db-parser pattern databases, where a name-value pair contains type-hint like syntax.

This PR is a call for discussion, it now introduces a v6 to the patterndb format and with older versions merely ignores the type-hint and assumes it's part of the value as is.

I am honestly torn between two alternatives, which I would need some feedback about:
  1) do as implemented in this PR, with version < 6, we just use template strings as they were, ignoring type-hints altogether
  2) use a separate "type" attribute in the <value> tag

Alternative 1.

## patterndb v5 (and before) would ignore type-hints in value tags
```
<patterndb version="5">
  <ruleset>
     <rule>
     <values>
        <value name="foobar">su(pam_unix)</value>
    </values>
  </ruleset>
</patterndb>
```

## patterndb v6 would allow type-hints or a string() if the value is otherwise conflicting
```
<patterndb version="6">
  <ruleset>
     <rule>
     <values>
        <value name="foobar">string(su(pam_unix))</value>
    </values>
  </ruleset>
</patterndb>
```

Alternative 2.

```
<patterndb version="5">
  <ruleset>
     <rule>
     <values>
        <value name="number" type="integer">42</value>
        <value name="foobar" type="string">su(pam_unix)</value>
    </values>
  </ruleset>
</patterndb>
```

With the 2nd alternative, we would probably still need to introduce v6 so that we document the "type" attribute, but otherwise the parser would remain the same.

@faxm0dem

erempel commented:
There are only a few places in the patterndb where types have any meaning.
1. tags
2. values
3. action conditions
4. possibly the test_values would require a type to correctly compare parsed values which get assigned typed values.

All other parts of the patterndb need to remain as strings.
1. ruleset name
2. ruleset pattern
3. rule_id
4. pattern
5. test_message

At the very least, places that can only ever have a string type should not have type hints applied.
The "XML way" for metadata is to include it as an attribute tag.
The use case that may not be covered by this is the action conditions. Action conditions can be multi-component expressions and may require more than one type hint and the hint will apply to only portions of the action condition.

<action condition='int($(context-length)) == int($max)'>

I have not thought through all of the use cases for the action condition but it occurs to me that any values used in the action condition could/would already have been assigned with a type hint, meaning that type hints are not required in action conditions at all. If that is the case then staying with the XML attribute would be more consistent with XML syntax.

In my particular use case we create the XML patterndb from a database using off the shelf XML writing tools. Having to type hint all of the data (values, tags, patterns ...) before feeding it to the XML writing tools would feel unnatural versus applying an attribute for the XML object.

I think that the XML attributes is the correct solution.

One other thing to consider is if the test_values actually require hinting. If they are being compared to values parsed from the pattern, then all of the values are typed based on the parser. Requiring that the patterndb author type each of the tests the same as the patterns that parse them will create uneccessary mismatch errors.

bazsi commented:
I have now pushed another iteration where I've changed db-parser() behaviour to what you suggested. I introduced a separate "type" attribute to both <value> and <test_value> tags, neither is required

* the one in <value> defaults to propagating the type of the template expression (e.g. if the template is typed without an explicit type-hint, that will be considered)
* the one in <test_value> will default to "string", so if not specified, pdbtool validate would fail if the template expression produces a non-string value

This is a change from previous 4.0 behaviour, but I kind of agree with @erempel that this is the right way and I should have chosen this initially when I implemented type-hinting in patterndb in the first place.

bazsi commented:
I have set it to target 4.0, as this is an incompatible change compared to our previous  intentions with 4.0.

bazsi commented:
See a couple more notes inline:

> There are only a few places in the patterndb where types have any meaning.
> 
>     1. tags
> 
>     2. values
> 
>     3. action conditions
> 
>     4. possibly the test_values would require a type to correctly compare parsed values which get assigned typed values.
> 
> 
> All other parts of the patterndb need to remain as strings.
> 
>     1. ruleset name
> 
>     2. ruleset pattern
> 
>     3. rule_id
> 
>     4. pattern
> 
>     5. test_message
> 

Agreed, and this is exactly what happens. NOTE: tags are not typed in syslog-ng, they are just boolean "flags" associated with a message.


> 
> At the very least, places that can only ever have a string type should not have type hints applied. The "XML way" for metadata is to include it as an attribute tag. The use case that may not be covered by this is the action conditions. Action conditions can be multi-component expressions and may require more than one type hint and the hint will apply to only portions of the action condition.
> 
> <action condition='int($(context-length)) == int($max)'>
> 
> I have not thought through all of the use cases for the action condition but it occurs to me that any values used in the action condition could/would already have been assigned with a type hint, meaning that type hints are not required in action conditions at all. If that is the case then staying with the XML attribute would be more consistent with XML syntax.

Conditions usually just use existing types, exactly as you mention. With the latest "type-aware-comparison" changes, syslog-ng tries to do the right thing when mismatching values are presented in a conditional, which you can fix with explicit type hints.

> 
> In my particular use case we create the XML patterndb from a database using off the shelf XML writing tools. Having to type hint all of the data (values, tags, patterns ...) before feeding it to the XML writing tools would feel unnatural versus applying an attribute for the XML object.
> 
> I think that the XML attributes is the correct solution.

I did exactly that.

> 
> One other thing to consider is if the test_values actually require hinting. If they are being compared to values parsed from the pattern, then all of the values are typed based on the parser. Requiring that the patterndb author type each of the tests the same as the patterns that parse them will create uneccessary mismatch errors.

Right now we don't require the attribute, but if it's unspecified it becomes "string", which is then explicitly validated.

This is the XML I was testing with:

```
<patterndb version='6' pub_date='2021-03-15'>
  <ruleset name='patternize' id='d414c981-2452-9142-9346-db2026d4e1b6'>
    <rules>
      <rule id='condor_schedd' class='sequence' provider='patternize'>
        <patterns>
          <pattern>Cluster @NUMBER:integer@ has no Owner attribute.  Ignoring...</pattern>
        </patterns>
        <values>
          <value name="seq-complexity" type="double">0.05</value>
          <value name="seq-matches">su(pam_unix)</value>
	  <value name="seq-created" type="datetime">1669135810</value>
          <value name="seq-last-match"/>
        </values>
        <examples>
          <example>
            <test_message program="">Cluster 123 has no Owner attribute.  Ignoring...</test_message>
            <test_values>
              <test_value name="seq-complexity" type="double">0.05</test_value>
              <test_value name="seq-created" type="datetime">2021-12-22</test_value>
            </test_values>
          </example>
        </examples>
      </rule>
    </rules>
  </ruleset>
</patterndb>
```

and the sample output:
```
$ bin/pdbtool test -v --validate etc/faxmodem.pdb
Testing message: program='' message='Cluster 123 has no Owner attribute.  Ignoring...'
 Match name='.classifier.rule_id', value='condor_schedd', type='string', expected='condor_schedd'
 Match name='seq-complexity', value='0.05', type='double', expected='0.05'
 Match name='seq-created', value='1669135810', type='datetime', expected='1669135810'
```

As you can see the type information has been added to the output.

bazsi commented:
if this is merged, the NEWS.md file should be updated something like this patch:

```
diff --git a/NEWS.md b/NEWS.md
index 773e482e8..5a0d0da8c 100644
--- a/NEWS.md
+++ b/NEWS.md
@@ -166,10 +166,10 @@ documented in the list down below:
 
   * db-parser() type support: db-parser() gets support for type casts,
     <value> assignments within db-parser() rules can associate types with
-    values using the type-casting syntax, e.g.  <value
-    name=”foobar”>int32($PID)</value>.  The “int32” is a type-cast that
+    values using the "type" attribute, e.g.  `<value
+    name="foobar" type="integer">$PID</value>`.  The “integer” is a type-cast that
     associates $foobar with an integer type.  db-parser()’s internal parsers
-    (e.g.  @NUMBER@) will also associated type information with a name-value
+    (e.g.  @NUMBER@) will also associate type information with a name-value
     pair automatically.
 
   * add-contextual-data() type support: any new name-value pair that is
@@ -435,11 +435,6 @@ The framework consists of these changes:
     case of the value field of the add-contextual-data() database contains an
     expression that resembles the new type-hinting syntax: type(value).
 
-  * `db-parser()`: add compatibility warnings and update advise in case the
-    <value>  element in the pattern database contains an expression that
-    resembles the new type-hinting syntax: type(value).
-    ([#4158](https://github.com/syslog-ng/syslog-ng/pull/4158))
-
   * `syslog-ng --help` screen: the output for the --help command line option has
     included sample paths to various files that contained autoconf style
     directory references (e.g. ${prefix}/etc for instance). This is now fixed,
```

faxm0dem commented:
Sorry if I'm late to the game, but I agree on Alternative 2. that is the attribute choice :+1:

bazsi commented on modules/dbparser/pdbtool/pdbtool.c:
```diff
@@ -279,7 +279,7 @@ pdbtool_merge(int argc, char *argv[])
   g_date_set_time_t(&date, time (NULL));
 
-  buff = g_markup_printf_escaped("<?xml version='1.0' encoding='UTF-8'?>\n<patterndb version='4' pub_date='%04d-%02d-%02d'>",
+  buff = g_markup_printf_escaped("<?xml version='1.0' encoding='UTF-8'?>\n<patterndb version='6' pub_date='%04d-%02d-%02d'>",
```
I just did that as well.

bazsi commented:
rebased against master.

MrAnno reviewed (approved):
Thank you.
